Documentation/design-docs/diagnostics/diagnostics-tools-contract.md

```diff
@@ -0,0 +1,49 @@
+# Diagnostics Tools Contract
+<<<<<<< HEAD
+ 
+To perform its function, diagnostics tools need to make assumptions on how the compiler, runtime, and framework works. This was done in various obscure ways. Often, the dependency is taken in another code base such that breaking changes are detected only in runtime during integration tests, which is often too late. 
+ 
+The crux of the problem is that the contract is implicit. Therefore, we are making changes to the product so that we establish a formal and explicit contract between the components and the diagnostics tools. For example, certain kind of breaking changes can now be detected during build time. You are reading this probably because you have made a change that broke the contract. In the sequel of the document, we will talk about why your change is breaking and what can you do about it.
+ 
+## Overall Principle
+At a breakpoint when the debuggee stops. We can look at any memory freely in the debuggee. This is often the most convenient way to perform any debugging operation. Therefore, unlike other applications, debuggers are mostly interested in the fields of the debuggee, not the APIs. This is especially true in the AOT scenario where we do not have support for true function evaluation in the debuggee, yet.
+
+The debugger accesses the field through symbols, so this is mostly a no-op for the framework developers. But there are a few exceptions because the debugger is trying to make sense of those objects and give better diagnostics experience, for example:
+ 
+### Strings
+In a typical application, you have many strings. Strings are just another managed object such that we can inspect all the fields out of it. There is a character array, we can inspect the characters. But that would be an awful experience. The user will want to just look at the string. In order to do that, the debugger makes an assumption that `System.String._firstChar` represent the head of the character array and get the string out the debuggee memory and display as a string. 
+ 
+Now imagine what would happen if the field is renamed? The debugger will not be able to find the string, and it may fallback any other approach. That is fundamentally why we need to keep the fields unchanged. Not only the name and its existence, but also its meaning. If _firstChar is repurposed to mean something else, the debugger would output something wrong, which is arguably worse.
```

Comment:
All code identifiers throughout the document should be enclosed in "``" - `_firstChar` on this line.